File: Python/quickSort.py
import random

def quickSort(l):
  pivot = l[0]
  index = 0

  enum = enumerate(l)
  for i, item in enum:
    # smaller and to the right
    if item < pivot and i > index:
      l.pop(i)
      l.insert(0, item)
      index += 1
    
    # larger and to the left
    if item > pivot and i < index:
      l.pop(i)
      l.append(item)
      index -= 1
  
  right = quickSort(l[index + 1:]) if index != len(l) - 1 else []
  left = quickSort(l[:index]) if index != 0 else []
  
  return left + [ pivot ] + right
    

File: JavaScript/quickSort.js
function quickSort(arr){
  if(arr.length <= 1) return arr

  let index = 0
  let pivot = arr[index];

  for(let i = 1; i < arr.length; i++){
    if (arr[i] < pivot && i > index){
      let [ item ] = arr.splice(i, 1)
      arr.unshift(item)
      index++
    }
    else if (arr[i] > pivot && i < index){
      let [ item ] = arr.splice(i, 1)
      arr.push(item)
      index--
    }
  }
  
  const left = quickSort(arr.slice(0, index))
  const right = quickSort(arr.slice(index + 1))

  
  return [...left, pivot, ...right]
}

console.log(quickSort([5, 3, 7, 4, 6, 2, 8, 1]))
